_posts/2024-03-20-kraft-migration.md

```diff
@@ -0,0 +1,203 @@
+```
+
+All the possible metadata changes are encoded as specific event records in the metadata log topic in Apache Kafka.
+They are all described using JSON files (which are then used to automatically build the corresponding Java classes) you can find [here](https://github.com/apache/kafka/tree/trunk/metadata/src/main/resources/common/metadata).
+
+Removing an external system like ZooKeeper simplifies the overall architecture and removes the burden of operating an additional component.
+The scalability is also improved by reducing the load on the metadata store by using snapshots to avoid unbounded growth (like compacted topic).
+When there is a leadership change in the controller quorum, the new leader already has all the committed metadata records so the recovery is pretty fast.
+
+Another interesting aspect of using the KRaft mode is about the role that a single node can have within the cluster itself.
+By using the new `process.roles` configuration parameter, a Kafka node can act as a `broker`, a `controller` or perform both roles in "combined" mode
+A `broker` node is responsible for handing requests, including reads and writes on topics, from Kafka clients.
+A `controller` node takes part in the controller quorum and replicates the metadata log. At any time, one of the `controller` nodes is the quorum leader and the others are followers that can take over in case the current leader resigns if it's restarted or fails for example.
+
+Effectively, `broker` and `controller` are two different services running on the node within the JVM.
+When in "combined" mode, a node performs both roles with the corresponding responsibilities.
+Using the "combined" mode allows to reduce the number of nodes within the cluster compared to a ZooKeeper configuration.
+It is not recommended for critical use cases as this offers less isolation in case of failures.
+This mode is useful for small use cases, for example it allows setting up a Kafka development environment by starting a single node.
+
+### How to migrate from ZooKeeper to KRaft
+
+Today, most of the existing Kafka clusters are ZooKeeper-based.
+Users are interested in migrating to KRaft mode in order to overcome all the limitations we have talked about.
+While KRaft is production ready for a few releases already, it lacked important features (JBOD support is still in early access) so few new clusters were deployed in KRaft mode.
+
+Furthermore, ZooKeeper is already considered as deprecated and, with the Kafka project expecting to remove ZooKeeper support in an upcoming 4.0 release, expected in 2024, it is now time to consider migrating clusters from ZooKeeper to KRaft.
+
+There is a manual procedure, consisting of several phases, to run in order to migrate a cluster without any downtime.
+The following content doesn't want to be an exhaustive description of the migration procedure but more an overview of how it is supposed to work.
```

Comment:
```suggestion
The following content aims to provide an overview of how the migration procedure is supposed to work, rather than offering an exhaustive description.
```